src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -116,11 +117,13 @@ service_node_test_results quorum_cop::check_service_node(
         pulse_participation = proof.pulse_participation;
         timestamp_participation = proof.timestamp_participation;
         timesync_status = proof.timesync_status;
+        l2_height = proof.proof->l2_height;
     });
     std::chrono::seconds time_since_last_uptime_proof{std::time(nullptr) - timestamp};
 
     bool check_uptime_obligation = true;
     bool check_checkpoint_obligation = true;
+    bool check_l2_height = m_l2_height > netconf.L2_TRACKER_SAFE_BLOCKS;
```

Comment:
took the second stricter approach. I havent added an extra tolerance amount because it feels like the `get_safe_height` has the tolerance built into that? Also we are enforcing timestamps being correct for nodes.

If we want more tolerance then I'd advocate we increase `get_safe_height`

Reply:
> took the second stricter approach. I havent added an extra tolerance amount because it feels like the get_safe_height has the tolerance built into that?

`get_safe_height` only has 10s tolerance built-in; that's sufficient for block processing (if you are more than that, it's reasonable that you fail a pulse quorum), where it takes multiple, sustained failures to participate before you would be penalized.

> If we want more tolerance then I'd advocate we increase get_safe_height

`get_safe_height` isn't the appropriate buffer here: that one exists to allow nodes to be up to 10s out of date with a 60s update frequency with no possible ill effect on the network.  Increasing it means delaying event propagation into the oxen chain, which seems unrelated to proof tolerance.

As currently implemented, your L2 provider being just 11s behind the tip of the chain even just momentarily at just the wrong moment (i.e. when you are sending a proof) means you produce a proof that will get you decommed at any point over the next hour.  None of our other enforcements of obligations are even remotely as strict: lokinet/storage server must be continuously unreachable for an hour; you must have missed *two* consecutive proofs (which means you have been disconnected for at *least* 1h5min); checkpoint/pulse/timesync failures all require at least 5 failures (of the last 8 expected participations) before they trigger.

Having a perfectly reliable L2 providers configured as a backup won't even save you from this potential decomm without a buffer: we only even check backups about every 3 minutes (and only switch to a backup if the primary is more than 30s out of sync).

We need either to build a margin into the l2_height that we allow, or we need to base a failure on multiple consecutive failures and not just one bad value in one proof.

Reply:
Alright scrapped the `get_safe_height` idea and have the given them 1h buffer defined in its own config variable. This is also adjusted by the proof age.

I dont have too strong an opinion on how long the buffer should be.

Reply:
Just realised what you were trying to achieve with the std::max. Will fix in the morning when i come online